// DOM.File.FileReader/index.js
"use strict";
var $foreign = require("./foreign");
var Prelude = require("../Prelude");
var Control_Monad_Eff = require("../Control.Monad.Eff");
var DOM = require("../DOM");
var DOM_File_FileReader_ReadyState = require("../DOM.File.FileReader.ReadyState");
var DOM_File_Types = require("../DOM.File.Types");
var Data_Foreign = require("../Data.Foreign");
var Data_Maybe = require("../Data.Maybe");
var Partial_Unsafe = require("../Partial.Unsafe");
var Control_Bind = require("../Control.Bind");
var Data_Function = require("../Data.Function");
var Control_Applicative = require("../Control.Applicative");
var readyState = function (fr) {
    return function __do() {
        var v = $foreign.readyStateImpl(fr)();
        return Partial_Unsafe.unsafePartial(function (dictPartial) {
            return Data_Maybe.fromJust(dictPartial)(DOM_File_FileReader_ReadyState.toEnumReadyState(v));
        });
    };
};
module.exports = {
    readyState: readyState, 
    abort: $foreign.abort, 
    fileReader: $foreign.fileReader, 
    readAsArrayBuffer: $foreign.readAsArrayBuffer, 
    readAsDataURL: $foreign.readAsDataURL, 
    readAsText: $foreign.readAsText, 
    result: $foreign.result
};
