// DOM.Node.Element/index.js
"use strict";
var $foreign = require("./foreign");
var Prelude = require("../Prelude");
var Control_Monad_Eff = require("../Control.Monad.Eff");
var Data_Nullable = require("../Data.Nullable");
var DOM = require("../DOM");
var DOM_Node_Types = require("../DOM.Node.Types");
module.exports = {
    className: $foreign.className, 
    clientHeight: $foreign.clientHeight, 
    clientLeft: $foreign.clientLeft, 
    clientTop: $foreign.clientTop, 
    clientWidth: $foreign.clientWidth, 
    getAttribute: $foreign.getAttribute, 
    getElementsByClassName: $foreign.getElementsByClassName, 
    getElementsByTagName: $foreign.getElementsByTagName, 
    getElementsByTagNameNS: $foreign.getElementsByTagNameNS, 
    id: $foreign.id, 
    localName: $foreign.localName, 
    namespaceURI: $foreign.namespaceURI, 
    prefix: $foreign.prefix, 
    removeAttribute: $foreign.removeAttribute, 
    scrollHeight: $foreign.scrollHeight, 
    scrollLeft: $foreign.scrollLeft, 
    scrollTop: $foreign.scrollTop, 
    scrollWidth: $foreign.scrollWidth, 
    setAttribute: $foreign.setAttribute, 
    setClassName: $foreign.setClassName, 
    setId: $foreign.setId, 
    setScrollLeft: $foreign.setScrollLeft, 
    setScrollTop: $foreign.setScrollTop, 
    tagName: $foreign.tagName
};
