// Signal.DOM/index.js
"use strict";
var $foreign = require("./foreign");
var Control_Monad_Eff = require("../Control.Monad.Eff");
var Control_Monad_Eff_Timer = require("../Control.Monad.Eff.Timer");
var DOM = require("../DOM");
var Prelude = require("../Prelude");
var Signal = require("../Signal");
var Signal_Time = require("../Signal.Time");
var Control_Bind = require("../Control.Bind");
var Data_Function = require("../Data.Function");
var Control_Applicative = require("../Control.Applicative");

/**
 *  |A signal which contains the document window's current width and height.
 */
var windowDimensions = $foreign.windowDimensionsP(Signal.constant);

/**
 *  |A signal containing the current state of the touch device, as described by
 *  |the `Touch` record type.
 */
var touch = $foreign.touchP(Signal.constant);

/**
 *  |A signal which will be `true` when at least one finger is touching the
 *  |touch device, and `false` otherwise.
 */
var tap = function __do() {
    var v = touch();
    return Signal.flippedMap(Signal.functorSignal)(v)(function (t) {
        if (t.length === 0) {
            return false;
        };
        return true;
    });
};

/**
 *  |A signal containing the current mouse position.
 */
var mousePos = $foreign.mousePosP(Signal.constant);

/**
 *  |Creates a signal which will be `true` when the given mouse button is
 *  |pressed, and `false` when it's released.
 */
var mouseButton = $foreign.mouseButtonP(Signal.constant);

/**
 *  |Creates a signal which will be `true` when the key matching the given key
 *  |code is pressed, and `false` when it's released.
 */
var keyPressed = $foreign.keyPressedP(Signal.constant);

/**
 *  |A signal which yields the current time, as determined by `now`, on every
 *  |animation frame (see [https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame]).
 */
var animationFrame = $foreign.animationFrameP(Signal.constant)(Signal_Time.now);
module.exports = {
    animationFrame: animationFrame, 
    keyPressed: keyPressed, 
    mouseButton: mouseButton, 
    mousePos: mousePos, 
    tap: tap, 
    touch: touch, 
    windowDimensions: windowDimensions
};
