// Graphics.Drawing/index.js

/**
 *  | This module defines a type `Drawing` for creating vector graphics.
 */
"use strict";
var Prelude = require("../Prelude");
var Color = require("../Color");
var Control_Alt = require("../Control.Alt");
var Control_Monad_Eff = require("../Control.Monad.Eff");
var Data_Foldable = require("../Data.Foldable");
var Data_List = require("../Data.List");
var Data_Maybe = require("../Data.Maybe");
var Data_Monoid = require("../Data.Monoid");
var Graphics_Canvas = require("../Graphics.Canvas");
var Graphics_Drawing_Font = require("../Graphics.Drawing.Font");
var $$Math = require("../Math");
var Data_Semigroup = require("../Data.Semigroup");
var Data_List_Types = require("../Data.List.Types");
var Data_Eq = require("../Data.Eq");
var Data_HeytingAlgebra = require("../Data.HeytingAlgebra");
var Data_Functor = require("../Data.Functor");
var Control_Applicative = require("../Control.Applicative");
var Data_Unit = require("../Data.Unit");
var Control_Bind = require("../Control.Bind");
var Data_Function = require("../Data.Function");
var Data_Semiring = require("../Data.Semiring");
var Control_Semigroupoid = require("../Control.Semigroupoid");

/**
 *  | Encapsulates shadow settings etc.
 */
var Shadow = function (x) {
    return x;
};

/**
 *  | A single shape.
 */
var Path = (function () {
    function Path(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    Path.create = function (value0) {
        return function (value1) {
            return new Path(value0, value1);
        };
    };
    return Path;
})();

/**
 *  | A single shape.
 */
var Rectangle = (function () {
    function Rectangle(value0) {
        this.value0 = value0;
    };
    Rectangle.create = function (value0) {
        return new Rectangle(value0);
    };
    return Rectangle;
})();

/**
 *  | A single shape.
 */
var Circle = (function () {
    function Circle(value0) {
        this.value0 = value0;
    };
    Circle.create = function (value0) {
        return new Circle(value0);
    };
    return Circle;
})();

/**
 *  | A single shape.
 */
var Composite = (function () {
    function Composite(value0) {
        this.value0 = value0;
    };
    Composite.create = function (value0) {
        return new Composite(value0);
    };
    return Composite;
})();

/**
 *  | Encapsulates outline color etc.
 */
var OutlineStyle = function (x) {
    return x;
};

/**
 *  | Encapsulates fill color etc.
 */
var FillStyle = function (x) {
    return x;
};

/**
 *  | A vector `Drawing`.
 */
var Fill = (function () {
    function Fill(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    Fill.create = function (value0) {
        return function (value1) {
            return new Fill(value0, value1);
        };
    };
    return Fill;
})();

/**
 *  | A vector `Drawing`.
 */
var Outline = (function () {
    function Outline(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    Outline.create = function (value0) {
        return function (value1) {
            return new Outline(value0, value1);
        };
    };
    return Outline;
})();

/**
 *  | A vector `Drawing`.
 */
var Text = (function () {
    function Text(value0, value1, value2, value3, value4) {
        this.value0 = value0;
        this.value1 = value1;
        this.value2 = value2;
        this.value3 = value3;
        this.value4 = value4;
    };
    Text.create = function (value0) {
        return function (value1) {
            return function (value2) {
                return function (value3) {
                    return function (value4) {
                        return new Text(value0, value1, value2, value3, value4);
                    };
                };
            };
        };
    };
    return Text;
})();

/**
 *  | A vector `Drawing`.
 */
var Many = (function () {
    function Many(value0) {
        this.value0 = value0;
    };
    Many.create = function (value0) {
        return new Many(value0);
    };
    return Many;
})();

/**
 *  | A vector `Drawing`.
 */
var Scale = (function () {
    function Scale(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    Scale.create = function (value0) {
        return function (value1) {
            return new Scale(value0, value1);
        };
    };
    return Scale;
})();

/**
 *  | A vector `Drawing`.
 */
var Translate = (function () {
    function Translate(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    Translate.create = function (value0) {
        return function (value1) {
            return new Translate(value0, value1);
        };
    };
    return Translate;
})();

/**
 *  | A vector `Drawing`.
 */
var Rotate = (function () {
    function Rotate(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    Rotate.create = function (value0) {
        return function (value1) {
            return new Rotate(value0, value1);
        };
    };
    return Rotate;
})();

/**
 *  | A vector `Drawing`.
 */
var Clipped = (function () {
    function Clipped(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    Clipped.create = function (value0) {
        return function (value1) {
            return new Clipped(value0, value1);
        };
    };
    return Clipped;
})();

/**
 *  | A vector `Drawing`.
 */
var WithShadow = (function () {
    function WithShadow(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    WithShadow.create = function (value0) {
        return function (value1) {
            return new WithShadow(value0, value1);
        };
    };
    return WithShadow;
})();

/**
 *  | Apply a translation by providing the x and y distances.
 */
var translate = function (tx) {
    return function (ty) {
        return Translate.create({
            translateX: tx, 
            translateY: ty
        });
    };
};

/**
 *  | Render some text.
 */
var text = Text.create;

/**
 *  | Set the shadow blur.
 */
var shadowOffset = function (x) {
    return function (y) {
        return {
            color: Data_Maybe.Nothing.value, 
            blur: Data_Maybe.Nothing.value, 
            offset: new Data_Maybe.Just({
                x: x, 
                y: y
            })
        };
    };
};

/**
 *  | Set the shadow color.
 */
var shadowColor = function (c) {
    return {
        color: new Data_Maybe.Just(c), 
        blur: Data_Maybe.Nothing.value, 
        offset: Data_Maybe.Nothing.value
    };
};

/**
 *  | Set the shadow blur.
 */
var shadowBlur = function (b) {
    return {
        color: Data_Maybe.Nothing.value, 
        blur: new Data_Maybe.Just(b), 
        offset: Data_Maybe.Nothing.value
    };
};

/**
 *  | Apply a `Shadow` to a `Drawing`.
 */
var shadow = WithShadow.create;
var semigroupShape = new Data_Semigroup.Semigroup(function (v) {
    return function (v1) {
        if (v instanceof Composite) {
            return new Composite(Data_Semigroup.append(Data_List_Types.semigroupList)(v.value0)(Data_List.singleton(v1)));
        };
        if (v1 instanceof Composite) {
            return new Composite(new Data_List_Types.Cons(v, v1.value0));
        };
        return new Composite(new Data_List_Types.Cons(v, new Data_List_Types.Cons(v1, Data_List_Types.Nil.value)));
    };
});
var semigroupShadow = new Data_Semigroup.Semigroup(function (v) {
    return function (v1) {
        return {
            color: Control_Alt.alt(Data_Maybe.altMaybe)(v.color)(v1.color), 
            blur: Control_Alt.alt(Data_Maybe.altMaybe)(v.blur)(v1.blur), 
            offset: Control_Alt.alt(Data_Maybe.altMaybe)(v.offset)(v1.offset)
        };
    };
});
var semigroupOutlineStyle = new Data_Semigroup.Semigroup(function (v) {
    return function (v1) {
        return {
            color: Control_Alt.alt(Data_Maybe.altMaybe)(v.color)(v1.color), 
            lineWidth: Control_Alt.alt(Data_Maybe.altMaybe)(v.lineWidth)(v1.lineWidth)
        };
    };
});
var semigroupFillStyle = new Data_Semigroup.Semigroup(function (v) {
    return function (v1) {
        return {
            color: Control_Alt.alt(Data_Maybe.altMaybe)(v.color)(v1.color)
        };
    };
});
var semigroupDrawing = new Data_Semigroup.Semigroup(function (v) {
    return function (v1) {
        if (v instanceof Many) {
            return new Many(Data_Semigroup.append(Data_List_Types.semigroupList)(v.value0)(Data_List.singleton(v1)));
        };
        if (v1 instanceof Many) {
            return new Many(new Data_List_Types.Cons(v, v1.value0));
        };
        return new Many(new Data_List_Types.Cons(v, new Data_List_Types.Cons(v1, Data_List_Types.Nil.value)));
    };
});

/**
 *  | Apply a scale transformation by providing the x and y scale factors.
 */
var scale = function (sx) {
    return function (sy) {
        return Scale.create({
            scaleX: sx, 
            scaleY: sy
        });
    };
};

/**
 *  | Apply a rotation by providing the angle.
 */
var rotate = Rotate.create;

/**
 *  | Render a `Drawing` to a canvas.
 */
var render = function (ctx) {
    var renderShape = function (v) {
        if (v instanceof Path && v.value1 instanceof Data_List_Types.Nil) {
            return Control_Applicative.pure(Control_Monad_Eff.applicativeEff)(Data_Unit.unit);
        };
        if (v instanceof Path && v.value1 instanceof Data_List_Types.Cons) {
            return function __do() {
                Graphics_Canvas.moveTo(ctx)(v.value1.value0.x)(v.value1.value0.y)();
                Data_Foldable.for_(Control_Monad_Eff.applicativeEff)(Data_List_Types.foldableList)(v.value1.value1)(function (p1) {
                    return Graphics_Canvas.lineTo(ctx)(p1.x)(p1.y);
                })();
                return Control_Applicative.when(Control_Monad_Eff.applicativeEff)(v.value0)(Data_Functor["void"](Control_Monad_Eff.functorEff)(Graphics_Canvas.closePath(ctx)))();
            };
        };
        if (v instanceof Rectangle) {
            return Data_Functor["void"](Control_Monad_Eff.functorEff)(Graphics_Canvas.rect(ctx)(v.value0));
        };
        if (v instanceof Circle) {
            return Data_Functor["void"](Control_Monad_Eff.functorEff)(Graphics_Canvas.arc(ctx)({
                x: v.value0.x, 
                y: v.value0.y, 
                r: v.value0.r, 
                start: 0.0, 
                end: $$Math.pi * 2.0
            }));
        };
        if (v instanceof Composite) {
            return Data_Foldable.for_(Control_Monad_Eff.applicativeEff)(Data_List_Types.foldableList)(v.value0)(renderShape);
        };
        throw new Error("Failed pattern match at Graphics.Drawing line 309, column 3 - line 309, column 39: " + [ v.constructor.name ]);
    };
    var applyShadow = function (v) {
        return function __do() {
            Data_Foldable.for_(Control_Monad_Eff.applicativeEff)(Data_Foldable.foldableMaybe)(v.color)(function (color) {
                return Graphics_Canvas.setShadowColor(Color.cssStringHSLA(color))(ctx);
            })();
            Data_Foldable.for_(Control_Monad_Eff.applicativeEff)(Data_Foldable.foldableMaybe)(v.blur)(function (blur) {
                return Graphics_Canvas.setShadowBlur(blur)(ctx);
            })();
            return Data_Foldable.for_(Control_Monad_Eff.applicativeEff)(Data_Foldable.foldableMaybe)(v.offset)(function (offset) {
                return function __do() {
                    Graphics_Canvas.setShadowOffsetX(offset.x)(ctx)();
                    return Graphics_Canvas.setShadowOffsetY(offset.y)(ctx)();
                };
            })();
        };
    };
    var applyOutlineStyle = function (v) {
        return function __do() {
            Data_Foldable.for_(Control_Monad_Eff.applicativeEff)(Data_Foldable.foldableMaybe)(v.color)(function (color) {
                return Graphics_Canvas.setStrokeStyle(Color.cssStringHSLA(color))(ctx);
            })();
            return Data_Foldable.for_(Control_Monad_Eff.applicativeEff)(Data_Foldable.foldableMaybe)(v.lineWidth)(function (width) {
                return Graphics_Canvas.setLineWidth(width)(ctx);
            })();
        };
    };
    var applyFillStyle = function (v) {
        return Data_Foldable.for_(Control_Monad_Eff.applicativeEff)(Data_Foldable.foldableMaybe)(v.color)(function (color) {
            return Graphics_Canvas.setFillStyle(Color.cssStringHSLA(color))(ctx);
        });
    };
    var go = function (v) {
        if (v instanceof Fill) {
            return Data_Functor["void"](Control_Monad_Eff.functorEff)(Graphics_Canvas.withContext(ctx)(function __do() {
                applyFillStyle(v.value1)();
                return Graphics_Canvas.fillPath(ctx)(renderShape(v.value0))();
            }));
        };
        if (v instanceof Outline) {
            return Data_Functor["void"](Control_Monad_Eff.functorEff)(Graphics_Canvas.withContext(ctx)(function __do() {
                applyOutlineStyle(v.value1)();
                return Graphics_Canvas.strokePath(ctx)(renderShape(v.value0))();
            }));
        };
        if (v instanceof Many) {
            return Data_Foldable.for_(Control_Monad_Eff.applicativeEff)(Data_List_Types.foldableList)(v.value0)(go);
        };
        if (v instanceof Scale) {
            return Data_Functor["void"](Control_Monad_Eff.functorEff)(Graphics_Canvas.withContext(ctx)(function __do() {
                Graphics_Canvas.scale(v.value0)(ctx)();
                return go(v.value1)();
            }));
        };
        if (v instanceof Translate) {
            return Data_Functor["void"](Control_Monad_Eff.functorEff)(Graphics_Canvas.withContext(ctx)(function __do() {
                Graphics_Canvas.translate(v.value0)(ctx)();
                return go(v.value1)();
            }));
        };
        if (v instanceof Rotate) {
            return Data_Functor["void"](Control_Monad_Eff.functorEff)(Graphics_Canvas.withContext(ctx)(function __do() {
                Graphics_Canvas.rotate(v.value0)(ctx)();
                return go(v.value1)();
            }));
        };
        if (v instanceof Clipped) {
            return Data_Functor["void"](Control_Monad_Eff.functorEff)(Graphics_Canvas.withContext(ctx)(function __do() {
                renderShape(v.value0)();
                Graphics_Canvas.clip(ctx)();
                return go(v.value1)();
            }));
        };
        if (v instanceof WithShadow) {
            return Data_Functor["void"](Control_Monad_Eff.functorEff)(Graphics_Canvas.withContext(ctx)(function __do() {
                applyShadow(v.value0)();
                return go(v.value1)();
            }));
        };
        if (v instanceof Text) {
            return Data_Functor["void"](Control_Monad_Eff.functorEff)(Graphics_Canvas.withContext(ctx)(function __do() {
                Graphics_Canvas.setFont(Graphics_Drawing_Font.fontString(v.value0))(ctx)();
                applyFillStyle(v.value3)();
                return Graphics_Canvas.fillText(ctx)(v.value4)(v.value1)(v.value2)();
            }));
        };
        throw new Error("Failed pattern match at Graphics.Drawing line 259, column 1 - line 316, column 40: " + [ v.constructor.name ]);
    };
    return go;
};

/**
 *  | Create a rectangle from the left, top, width and height parameters.
 */
var rectangle = function (x) {
    return function (y) {
        return function (w) {
            return function (h) {
                return new Rectangle({
                    x: x, 
                    y: y, 
                    w: w, 
                    h: h
                });
            };
        };
    };
};

/**
 *  | Create a path.
 */
var path = function (dictFoldable) {
    return function ($152) {
        return Path.create(false)(Data_List.fromFoldable(dictFoldable)($152));
    };
};

/**
 *  | Draw the outline of a `Shape`.
 */
var outlined = Data_Function.flip(Outline.create);

/**
 *  | Set the outline color.
 */
var outlineColor = function (c) {
    return {
        color: new Data_Maybe.Just(c), 
        lineWidth: Data_Maybe.Nothing.value
    };
};
var monoidShape = new Data_Monoid.Monoid(function () {
    return semigroupShape;
}, new Composite(Data_Monoid.mempty(Data_List_Types.monoidList)));
var monoidShadow = new Data_Monoid.Monoid(function () {
    return semigroupShadow;
}, {
    color: Data_Maybe.Nothing.value, 
    blur: Data_Maybe.Nothing.value, 
    offset: Data_Maybe.Nothing.value
});
var monoidOutlineStyle = new Data_Monoid.Monoid(function () {
    return semigroupOutlineStyle;
}, {
    color: Data_Maybe.Nothing.value, 
    lineWidth: Data_Maybe.Nothing.value
});
var monoidFillStyle = new Data_Monoid.Monoid(function () {
    return semigroupFillStyle;
}, {
    color: Data_Maybe.Nothing.value
});
var monoidDrawing = new Data_Monoid.Monoid(function () {
    return semigroupDrawing;
}, new Many(Data_Monoid.mempty(Data_List_Types.monoidList)));

/**
 *  | Set the line width.
 */
var lineWidth = function (c) {
    return {
        color: Data_Maybe.Nothing.value, 
        lineWidth: new Data_Maybe.Just(c)
    };
};

/**
 *  | Fill a `Shape`.
 */
var filled = Data_Function.flip(Fill.create);

/**
 *  | Set the fill color.
 */
var fillColor = function (c) {
    return {
        color: new Data_Maybe.Just(c)
    };
};

/**
 *  | Modify a `Drawing` by applying a transformation to every subdrawing.
 */
var everywhere = function (f) {
    var go = function (v) {
        if (v instanceof Many) {
            return f(new Many(Data_Functor.map(Data_List_Types.functorList)(go)(v.value0)));
        };
        if (v instanceof Scale) {
            return f(new Scale(v.value0, go(v.value1)));
        };
        if (v instanceof Translate) {
            return f(new Translate(v.value0, go(v.value1)));
        };
        if (v instanceof Rotate) {
            return f(new Rotate(v.value0, go(v.value1)));
        };
        if (v instanceof Clipped) {
            return f(new Clipped(v.value0, go(v.value1)));
        };
        if (v instanceof WithShadow) {
            return f(new WithShadow(v.value0, go(v.value1)));
        };
        return f(v);
    };
    return go;
};
var eqShape = new Data_Eq.Eq(function (v) {
    return function (v1) {
        if (v instanceof Path && v1 instanceof Path) {
            return v.value0 === v1.value0 && (Data_Eq.eq(Data_List_Types.eqList(Data_Eq.eqNumber))(Data_Functor.map(Data_List_Types.functorList)(function (v2) {
                return v2.x;
            })(v.value1))(Data_Functor.map(Data_List_Types.functorList)(function (v2) {
                return v2.x;
            })(v1.value1)) && Data_Eq.eq(Data_List_Types.eqList(Data_Eq.eqNumber))(Data_Functor.map(Data_List_Types.functorList)(function (v2) {
                return v2.y;
            })(v.value1))(Data_Functor.map(Data_List_Types.functorList)(function (v2) {
                return v2.y;
            })(v1.value1)));
        };
        if (v instanceof Rectangle && v1 instanceof Rectangle) {
            return v.value0.x === v1.value0.x && (v.value0.y === v1.value0.y && (v.value0.w === v1.value0.w && v.value0.h === v1.value0.h));
        };
        if (v instanceof Circle && v1 instanceof Circle) {
            return v.value0.x === v1.value0.x && (v.value0.y === v1.value0.y && v.value0.r === v1.value0.r);
        };
        if (v instanceof Composite && v1 instanceof Composite) {
            return Data_Eq.eq(Data_List_Types.eqList(eqShape))(v.value0)(v1.value0);
        };
        return false;
    };
});
var eqShadow = new Data_Eq.Eq(function (v) {
    return function (v1) {
        return Data_Eq.eq(Data_Maybe.eqMaybe(Color.eqColor))(v.color)(v1.color) && (Data_Eq.eq(Data_Maybe.eqMaybe(Data_Eq.eqNumber))(v.blur)(v1.blur) && Data_Maybe.maybe(Data_Maybe.isNothing(v1.offset))(function (o) {
            return Data_Maybe.maybe(false)(function (o$prime) {
                return o.x === o$prime.x && o.y === o$prime.y;
            })(v1.offset);
        })(v.offset));
    };
});
var eqOutlineStyle = new Data_Eq.Eq(function (v) {
    return function (v1) {
        return Data_Eq.eq(Data_Maybe.eqMaybe(Color.eqColor))(v.color)(v1.color) && Data_Eq.eq(Data_Maybe.eqMaybe(Data_Eq.eqNumber))(v.lineWidth)(v1.lineWidth);
    };
});
var eqFillStyle = new Data_Eq.Eq(function (v) {
    return function (v1) {
        return Data_Eq.eq(Data_Maybe.eqMaybe(Color.eqColor))(v.color)(v1.color);
    };
});
var eqDrawing = new Data_Eq.Eq(function (v) {
    return function (v1) {
        if (v instanceof Fill && v1 instanceof Fill) {
            return Data_Eq.eq(eqShape)(v.value0)(v1.value0) && Data_Eq.eq(eqFillStyle)(v.value1)(v1.value1);
        };
        if (v instanceof Outline && v1 instanceof Outline) {
            return Data_Eq.eq(eqShape)(v.value0)(v1.value0) && Data_Eq.eq(eqOutlineStyle)(v.value1)(v1.value1);
        };
        if (v instanceof Text && v1 instanceof Text) {
            return Data_Eq.eq(Graphics_Drawing_Font.eqFont)(v.value0)(v1.value0) && (v.value1 === v1.value1 && (v.value2 === v1.value2 && (Data_Eq.eq(eqFillStyle)(v.value3)(v1.value3) && v.value4 === v1.value4)));
        };
        if (v instanceof Many && v1 instanceof Many) {
            return Data_Eq.eq(Data_List_Types.eqList(eqDrawing))(v.value0)(v1.value0);
        };
        if (v instanceof Scale && v1 instanceof Scale) {
            return v.value0.scaleX === v1.value0.scaleX && (v.value0.scaleY === v1.value0.scaleY && Data_Eq.eq(eqDrawing)(v.value1)(v1.value1));
        };
        if (v instanceof Translate && v1 instanceof Translate) {
            return v.value0.translateX === v1.value0.translateX && (v.value0.translateY === v1.value0.translateY && Data_Eq.eq(eqDrawing)(v.value1)(v1.value1));
        };
        if (v instanceof Rotate && v1 instanceof Rotate) {
            return v.value0 === v1.value0 && Data_Eq.eq(eqDrawing)(v.value1)(v1.value1);
        };
        if (v instanceof Clipped && v1 instanceof Clipped) {
            return Data_Eq.eq(eqShape)(v.value0)(v1.value0) && Data_Eq.eq(eqDrawing)(v.value1)(v1.value1);
        };
        if (v instanceof WithShadow && v1 instanceof WithShadow) {
            return Data_Eq.eq(eqShadow)(v.value0)(v1.value0) && Data_Eq.eq(eqDrawing)(v.value1)(v1.value1);
        };
        return false;
    };
});

/**
 *  | Create a _closed_ path.
 */
var closed = function (dictFoldable) {
    return function ($153) {
        return Path.create(true)(Data_List.fromFoldable(dictFoldable)($153));
    };
};

/**
 *  | Clip a `Drawing` to a `Shape`.
 */
var clipped = Clipped.create;

/**
 *  | Create a circle from the left, top and radius parameters.
 */
var circle = function (x) {
    return function (y) {
        return function (r) {
            return new Circle({
                x: x, 
                y: y, 
                r: r
            });
        };
    };
};
module.exports = {
    circle: circle, 
    clipped: clipped, 
    closed: closed, 
    everywhere: everywhere, 
    fillColor: fillColor, 
    filled: filled, 
    lineWidth: lineWidth, 
    outlineColor: outlineColor, 
    outlined: outlined, 
    path: path, 
    rectangle: rectangle, 
    render: render, 
    rotate: rotate, 
    scale: scale, 
    shadow: shadow, 
    shadowBlur: shadowBlur, 
    shadowColor: shadowColor, 
    shadowOffset: shadowOffset, 
    text: text, 
    translate: translate, 
    semigroupShape: semigroupShape, 
    monoidShape: monoidShape, 
    eqShape: eqShape, 
    semigroupFillStyle: semigroupFillStyle, 
    monoidFillStyle: monoidFillStyle, 
    eqFillStyle: eqFillStyle, 
    semigroupOutlineStyle: semigroupOutlineStyle, 
    monoidOutlineStyle: monoidOutlineStyle, 
    eqOutlineStyle: eqOutlineStyle, 
    eqShadow: eqShadow, 
    semigroupShadow: semigroupShadow, 
    monoidShadow: monoidShadow, 
    semigroupDrawing: semigroupDrawing, 
    monoidDrawing: monoidDrawing, 
    eqDrawing: eqDrawing
};
